compiler/rustc_const_eval/src/interpret/projection.rs

```diff
@@ -203,23 +202,25 @@ where
 
     /// Iterates over all fields of an array. Much more efficient than doing the
     /// same by repeatedly calling `operand_index`.
-    pub fn project_array_fields<'a, P: Projectable<'tcx, M::Provenance>>(
+    pub fn project_array_fields<P: Projectable<'tcx, M::Provenance>>(
         &self,
-        base: &'a P,
-    ) -> InterpResult<'tcx, impl Iterator<Item = InterpResult<'tcx, P>> + 'a>
-    where
-        'tcx: 'a,
-    {
+        base: &P,
+    ) -> InterpResult<'tcx, impl Iterator<Item = InterpResult<'tcx, P>>> {
         let abi::FieldsShape::Array { stride, .. } = base.layout().fields else {
             span_bug!(self.cur_span(), "operand_array_fields: expected an array layout");
         };
         let len = base.len(self)?;
         let field_layout = base.layout().field(self, 0);
-        let tcx: TyCtxt<'tcx> = *self.tcx;
-        // `Size` multiplication
-        Ok((0..len).map(move |i| {
-            base.offset_with_meta(stride * i, MemPlaceMeta::None, field_layout, &tcx)
-        }))
+        // Due to lifetime issues, we have to buffer the fields in a `Vec`:
+        // We cannot keep borrowing `self` during iteration (users will want to mutate `InterpCx` during iteration),
+        // and yet we need `self` for `offset` (because it needs to do a bounds check).
+        let fields: Vec<_> = (0..len)
+            .map(move |i| {
+                // `Size` multiplication
+                base.offset(stride * i, field_layout, self)
+            })
+            .collect();
```

Comment:
If this is too bad for perf, we could in principle do a single bounds-check upfront and then use non-bounds-checked arithmetic -- but that would require poking some holes into the `Projectable` abstraction so I'd rather not do that...